tests/plugins/Qukeys/basic/test/testcase.cpp

```diff
@@ -0,0 +1,164 @@
+/* -*- mode: c++ -*-
+ * Copyright (C) 2020  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "testing/setup-googletest.h"
+
+#include "../sketch.h"
+
+SETUP_GOOGLETEST();
+
+namespace kaleidoscope {
+namespace testing {
+namespace {
+
+struct QukeyValues {
+  Key primary{Key_NoKey};
+  Key alternate{Key_NoKey};
+};
+constexpr KeyAddr key_addr_A{2, 1};
+constexpr KeyAddr key_addr_S{2, 2};
+constexpr KeyAddr key_addr_D{2, 3};
+constexpr KeyAddr key_addr_F{2, 4};
+constexpr KeyAddr key_addr_X{3, 2};
+
+using ::testing::IsEmpty;
+
+class QukeysBasic : public VirtualDeviceTest {};
+
+TEST_F(QukeysBasic, TapQukeyAlone) {
+
+  std::unique_ptr<State> state{nullptr};
+  std::set<uint8_t> expected_keycodes{};
+
+  // Press `A`
+  sim_.Press(key_addr_A);
+
+  state = VirtualDeviceTest::RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 0);
+
+  // Release `A`
+  sim_.Release(key_addr_A);
+  expected_keycodes.insert(Key_A.getKeyCode());
+
+  state = VirtualDeviceTest::RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+              ::testing::ElementsAreArray(expected_keycodes));
+
+  expected_keycodes.erase(Key_A.getKeyCode());
+
+  for (int i{0}; i < 10; ++i) {
```

Comment:
Comments around what the tests are testing would be very useful here. One way you can do that is to add them as strings after the assertions.

For example:

EXPECT_THAT(foo, ::testing::Le(2)) << "We still generated key reports for every cycle after the second one";

As an aside, I totally don't understand the logic of what's going on in this loop.

Reply:
Thanks for the advice. As it turns out, I had already stumbled across that, and have it on my list of improvements to make.

Regarding what's going on in the loop: In the current version of Qukeys, a tap results in two reports after the key is released, and because of the mid-cycle report problem, I made sure to prevent it from sending more than one report per cycle. That mid-cycle report problem will be solved by the event-driven code that I'm working on in another branch, and I don't want the test to fail because it sends two reports in one cycle.

Now, as it turns out, when I wrote this little loop, I was still unaware of how the test simulator accumulates reports (and purges them), so I could simplify it quite a bit. It's unnecessarily complex because I thought calling `sim_.RunCycle()` twice in a row would throw away any information about a report generated by the first cycle. What the loop is doing is checking up to ten times for a report. It's totally unnecessary.